docs/torch/ray_usage_advanced.rst

```diff
@@ -0,0 +1,247 @@
+    ...     pipe = dataset.window(blocks_per_window=10).repeat(config["n_epochs"])  # -> DatasetPipeline(num_windows=1, num_stages=1)
+    ...     q = GATQuery()
+    ...     def transform_batch(batch: list) -> dict:
+    ...         return q.query(g, batch)  # When we reference the server g in transform, it uses Client instead
+    ...     pipe = pipe.map_batches(transform_batch)
+    ...
+    ...     # Execute the training loop
+    ...     model.train()
+    ...     for epoch, epoch_pipe in enumerate(pipe.iter_epochs()):
+    ...         epoch_pipe = epoch_pipe.random_shuffle_each_window()
+    ...         for i, batch in enumerate(epoch_pipe.iter_torch_batches(batch_size=config["batch_size"])):
+    ...             scores = model(batch)
+    ...             labels = batch["labels"][batch["input_mask"]].flatten()
+    ...             loss = loss_fn(scores.type(torch.float32), labels)
+    ...             optimizer.zero_grad()
+    ...             loss.backward()
+    ...             optimizer.step()
+    ...
+    ...             session.report({"metric": (scores.argmax(1) == labels).sum(), "loss": loss.item()})
+    ...
+    ...     torch.save(model.state_dict(), config["model_dir"])
+
+In this step we start the training job.
+First we start a local ray cluster with `ray.init() <https://docs.ray.io/en/latest/ray-core/package-ref.html#ray-init>`.
+Next we initialize a `TorchTrainer <https://docs.ray.io/en/latest/ray-air/package-ref.html#pytorch>`
+object to wrap our training loop. This takes parameters that go to the training loop and parameters
+to define number workers and cpus/gpus used.
+Finally we call trainer.fit() to execute the training loop.
+
+.. code-block:: python
+
+    >>> model_dir = tempfile.TemporaryDirectory()
+
+    >>> ray.init()
+    RayContext(...)
+    >>> trainer = TorchTrainer(
+    ...     train_func,
+    ...     train_loop_config={
+    ...         "batch_size": 2708,
+    ...         "data_dir": data_dir.name,
+    ...         "sample_filename": "train.nodes",
+    ...         "n_epochs": 1,
+    ...         "model_dir": f"{model_dir.name}/model.pt",
+    ...     },
+    ...     run_config=RunConfig(verbose=0),
+    ...     scaling_config=ScalingConfig(num_workers=1, use_gpu=False, _max_cpu_fraction_per_node = 0.8),
+    ... )
+    >>> result = trainer.fit()
+
+Evaluate
+========
+
+.. code-block:: python
+
+    >>> trainer = TorchTrainer(
+    ...     train_func,
+    ...     train_loop_config={
+    ...         "batch_size": 2708,
+    ...         "data_dir": data_dir.name,
+    ...         "sample_filename": "test.nodes",
+    ...         "n_epochs": 1,
+    ...         "model_dir": f"{model_dir.name}/model.pt",
+    ...     },
+    ...     run_config=RunConfig(verbose=0),
+    ...     scaling_config=ScalingConfig(num_workers=1, use_gpu=False, _max_cpu_fraction_per_node = 0.8),
+    ... )
+    >>> result = trainer.fit()
```

Comment:
what is the expected result, how do we know training produced a meaningful model?

Reply:
This will be the same as in the #1, Right now I am printing metric and loss.